File: packages/shared-types/src/generation.ts
/**
 * A resolved citation traceable to an exact transcript chunk and timestamp boundary.
 */
export interface Citation {
  chunkId: string;
  start: number;
  end: number;
  score?: number;
}

/**
 * Structured answer returned by the grounded generation engine.
 */
export interface GroundedAnswer {
  text: string;
  grounded: boolean;
  citations: Citation[];
}

/**
 * Request payload to ask a question about a video's transcript.
 */
export interface AskQuestionRequest {
  question: string;
  conversationId?: string;
  clientRequestId?: string;
  topK?: number;
}

/**
 * Response envelope returned upon successful question answering.
 */
export interface AskQuestionResponse {
  success: true;
  videoId: string;
  conversationId?: string;
  answer: GroundedAnswer;
  message?: {
    id: string;
    role: "assistant";
    text: string;
    grounded: boolean;
    citations: Citation[];
    createdAt?: string;
  };
  requestId: string;
}

/**
 * Domain error codes specific to grounded answer generation.
 */
export type GenerationErrorCode =
  | "INVALID_QUESTION"
  | "GENERATION_FAILED"
  | "LLM_PROVIDER_UNAVAILABLE"
  | "LLM_TIMEOUT"
  | "LLM_OUTPUT_INVALID";

File: apps/backend/src/schemas/generation.py
"""Generation data models and schemas."""

from typing import Any, Literal

from src.schemas.base import CamelModel


class Citation(CamelModel):
    chunk_id: str
    start: float
    end: float
    score: float | None = None


class GroundedAnswer(CamelModel):
    text: str
    grounded: bool
    citations: list[Citation]
    message_id: str | None = None
    model: str = "gemini-1.5-flash"


class AssistantMessageSummary(CamelModel):
    id: str
    role: Literal["assistant"] = "assistant"
    text: str
    grounded: bool
    citations: list[Citation]
    created_at: str | None = None


class AskQuestionRequest(CamelModel):
    question: str
    conversation_id: str | None = None
    client_request_id: str | None = None
    top_k: int = 5


class AskQuestionResponse(CamelModel):
    success: bool = True
    video_id: str
    conversation_id: str | None = None
    answer: GroundedAnswer
    message: AssistantMessageSummary | None = None
    request_id: str | None = None


# Typed Streaming Protocol (SSE)
class StreamStartEvent(CamelModel):
    type: Literal["start"] = "start"
    request_id: str
    conversation_id: str
    message_id: str


class StreamTokenEvent(CamelModel):
    type: Literal["token"] = "token"
    text: str


class StreamCitationEvent(CamelModel):
    type: Literal["citation"] = "citation"
    citation: Citation


class StreamDoneEvent(CamelModel):
    type: Literal["done"] = "done"
    message: Any  # ConversationMessage


class StreamErrorEvent(CamelModel):
    type: Literal["error"] = "error"
    code: str
    message: str


StreamEvent = (
    StreamStartEvent | StreamTokenEvent | StreamCitationEvent | StreamDoneEvent | StreamErrorEvent
)

# Aliases for compatibility
StartStreamEvent = StreamStartEvent
TokenStreamEvent = StreamTokenEvent
CitationStreamEvent = StreamCitationEvent
DoneStreamEvent = StreamDoneEvent
ErrorStreamEvent = StreamErrorEvent
